### PR TITLE
#430 avoid use on-heap BitSet in ColumnValues

### LINKED ISSUE #430: In ColumnValues, BitSet should be in off-heap memory
Use Spark BitSet means copy data to on-heap memory. Lets use off-heap memory directly.

### PR DESCRIPTION
**Pull request by adrian-wang**
## What changes were proposed in this pull request?

Use Spark BitSet means copy data to on-heap memory. Lets use off-heap memory directly.
This resolves #430 

## How was this patch tested?

current tests.

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 1]
```diff
@@ -31,19 +31,13 @@ class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache)
   // for any FiberData, the first defaultSize / 8 will be the bitmask
   // TODO what if defaultSize / 8 is not an integer?
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: Do we need to add a `require(defaultSize % 8 == 0)`?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 2]
```diff
@@ -31,19 +31,13 @@ class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache)
...
-  // TODO get the bitset from the FiberByteData
-  val bitset: BitSet = {
-    val bs = new BitSet(defaultSize)
-    val longs = bs.toLongArray()
-    buffer.copyMemoryToLongs(0, longs)
+  private val dataOffset = (((defaultSize - 1) >> 6) + 1) * 8
 
-    bs
+  def isNullAt(idx: Int): Boolean = {
+    val bitmask = 1L << (idx & 0x3f)   // mod 64 and shift
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: `bitmask` is Int or Long here?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 3]
```diff
@@ -31,19 +31,13 @@ class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache)
   // for any FiberData, the first defaultSize / 8 will be the bitmask
   // TODO what if defaultSize / 8 is not an integer?
 
-  // TODO get the bitset from the FiberByteData
-  val bitset: BitSet = {
-    val bs = new BitSet(defaultSize)
-    val longs = bs.toLongArray()
-    buffer.copyMemoryToLongs(0, longs)
+  private val dataOffset = (((defaultSize - 1) >> 6) + 1) * 8
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: what's the difference with:
`dataOffset = ((defaultSize - 1) >> 3) + 1`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala (hunk 1 above)
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: `defaultSize` is not guranteed as 8 mod, so?

Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: In `DataFile` when we create `ColumnValues`, the `rowCount` can be any positive Integer?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 4]
```diff
@@ -31,19 +31,13 @@ class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache)
   // for any FiberData, the first defaultSize / 8 will be the bitmask
   // TODO what if defaultSize / 8 is not an integer?
 
-  // TODO get the bitset from the FiberByteData
-  val bitset: BitSet = {
-    val bs = new BitSet(defaultSize)
-    val longs = bs.toLongArray()
-    buffer.copyMemoryToLongs(0, longs)
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: `copyMemoryToLongs` in `MemoryManager` can also be removed?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala (hunk 3 above)
Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: Why is `(((defaultSize - 1) >> 6) + 1) * 8`?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 5]
```diff
@@ -31,19 +31,13 @@ class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache)
...
+  private val dataOffset = (((defaultSize - 1) >> 6) + 1) * 8
 
-    bs
+  def isNullAt(idx: Int): Boolean = {
+    val bitmask = 1L << (idx & 0x3f)   // mod 64 and shift
+    (buffer.getLong((idx >> 6) * 8) & bitmask) == 0  // div by 64 and mask
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: `>> 6 * 8`?? `>> 3`?

Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: Also, can we add some comments to explain with a little more details?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala (hunk 4 above)
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: I'll remove that in another pull request, which will also avoid other use of memory copy.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 6]
```diff
@@ -28,22 +28,15 @@ import org.apache.spark.util.collection.BitSet
 class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache) {
   require(dataType.isInstanceOf[AtomicType], s"Only atomic type accepted for now, got $dataType.")
 
-  // for any FiberData, the first defaultSize / 8 will be the bitmask
-  // TODO what if defaultSize / 8 is not an integer?
+  // for any FiberData, the first `(defaultSize - 1) >> 6 + 1` longs will be the bitmask
+  // num of bytes needed to hold defaultSize elements is then `((defaultSize - 1) >> 3) + 8`
+  private val dataOffset = ((defaultSize - 1) >> 6 << 3) + 8
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: :nit `>>3`

Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: That's not the same.
For example, `0x0FFF >> 6 << 3` is `0x01F8`, while `>>3` is `0x01FF`

Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: Right : ) Then the comments?

**Pull request merged**